File: android/src/main/kotlin/com/mithra/flutter/sdk/NaryaConfigurationCodec.kt
package com.mithra.flutter.sdk

import android.app.Application
import com.mithra.sdk.kotlin.android.Configuration
import com.mithra.sdk.kotlin.android.SessionConfiguration
import com.mithra.sdk.kotlin.core.InAppColorScheme
import com.mithra.sdk.kotlin.core.InAppConfiguration
import com.mithra.sdk.kotlin.core.NaryaEnvironment
import com.mithra.sdk.kotlin.core.internals.logger.Logger

/** Thrown when the Dart configuration map cannot produce a native configuration. */
internal class NaryaConfigurationException(
    val code: String,
    override val message: String,
) : RuntimeException(message)

/**
 * Decodes the `NaryaConfiguration.toMap()` wire form into the native
 * [Configuration].
 *
 * Defaults intentionally mirror the Dart defaults so an absent key and an
 * explicit default behave identically.
 */
internal object NaryaConfigurationCodec {

    fun decode(application: Application, source: Map<*, *>?): Configuration {
        if (source == null) {
            throw NaryaConfigurationException(
                "invalid_argument",
                "initialize was called without a configuration.",
            )
        }
        val writeKey = (source["writeKey"] as? String)?.takeIf { it.isNotBlank() }
            ?: throw NaryaConfigurationException(
                "missing_write_key",
                "NaryaConfiguration.writeKey is required and must not be blank.",
            )

        val environment = when (source["environment"] as? String) {
            "staging" -> NaryaEnvironment.STAGING
            else -> NaryaEnvironment.PRODUCTION
        }
        val dataPlaneUrl = (source["dataPlaneUrl"] as? String)?.takeIf { it.isNotBlank() }

        return Configuration(
            application = application,
            writeKey = writeKey,
            environment = environment,
            dataPlaneUrl = dataPlaneUrl ?: environment.dataPlaneUrl,
            trackApplicationLifecycleEvents =
                source["trackApplicationLifecycleEvents"] as? Boolean ?: true,
            trackDeepLinks = source["trackDeepLinks"] as? Boolean ?: true,
            // Flutter screen tracking is NaryaRouteObserver; native activity
            // tracking would emit one event for the single FlutterActivity.
            trackActivities = false,
            collectDeviceId = source["collectDeviceId"] as? Boolean ?: true,
            gzipEnabled = source["gzipEnabled"] as? Boolean ?: false,
            sessionConfiguration = decodeSession(source["session"] as? Map<*, *>),
            logLevel = decodeLogLevel(source["logLevel"] as? String),
            inApp = decodeInApp(source["inApp"] as? Map<*, *>),
        )
    }

    private fun decodeSession(source: Map<*, *>?): SessionConfiguration {
        val defaults = SessionConfiguration()
        if (source == null) return defaults
        return SessionConfiguration(
            automaticSessionTracking = source["automaticSessionTracking"] as? Boolean
                ?: defaults.automaticSessionTracking,
            sessionTimeoutInMillis = (source["sessionTimeoutInMillis"] as? Number)?.toLong()
                ?: defaults.sessionTimeoutInMillis,
            updateSessionOnBackgroundEvents =
                source["updateSessionOnBackgroundEvents"] as? Boolean
                    ?: defaults.updateSessionOnBackgroundEvents,
        )
    }

    private fun decodeInApp(source: Map<*, *>?): InAppConfiguration {
        val defaults = InAppConfiguration()
        if (source == null) return defaults
        return InAppConfiguration(
            enabled = source["enabled"] as? Boolean ?: defaults.enabled,
            displayInterval = (source["displayIntervalInSeconds"] as? Number)?.toDouble()
                ?: defaults.displayInterval,
            autoDisplayPaused = source["autoDisplayPaused"] as? Boolean
                ?: defaults.autoDisplayPaused,
            useInMemoryStorage = source["useInMemoryStorage"] as? Boolean
                ?: defaults.useInMemoryStorage,
            colorScheme = when (source["colorScheme"] as? String) {
                "light" -> InAppColorScheme.LIGHT
                "dark" -> InAppColorScheme.DARK
                "system" -> InAppColorScheme.SYSTEM
                else -> defaults.colorScheme
            },
            webViewBaseUrl = (source["webViewBaseUrl"] as? String)?.takeIf { it.isNotBlank() },
        )
    }

    private fun decodeLogLevel(source: String?): Logger.LogLevel = when (source) {
        "verbose" -> Logger.LogLevel.VERBOSE
        "debug" -> Logger.LogLevel.DEBUG
        "info" -> Logger.LogLevel.INFO
        "warn" -> Logger.LogLevel.WARN
        "error" -> Logger.LogLevel.ERROR
        else -> Logger.LogLevel.NONE
    }

    /**
     * A comparison key used to decide whether a repeated `initialize` call is
     * an idempotent no-op or a genuine reconfiguration attempt.
     *
     * The key is canonical - keys sorted at every nesting level - so a map that
     * arrived over the platform channel and the same map restored from the
     * persisted JSON by [NaryaAnalyticsHolder] compare equal even though their
     * concrete map types iterate in a different order.
     */
    fun identityOf(source: Map<*, *>?): String = canonical(source)

    private fun canonical(value: Any?): String = when (value) {
        null -> "null"
        is Map<*, *> -> value.entries
            .sortedBy { it.key?.toString() ?: "" }
            .joinToString(separator = "|", prefix = "{", postfix = "}") {
                "${it.key}=${canonical(it.value)}"
            }
        is Iterable<*> -> value.joinToString(separator = ",", prefix = "[", postfix = "]") {
            canonical(it)
        }
        is Number -> canonicalNumber(value)
        else -> value.toString()
    }

    /** `30` and `30.0` denote the same setting whichever numeric type decoded them. */
    private fun canonicalNumber(value: Number): String {
        val asDouble = value.toDouble()
        return if (asDouble == Math.floor(asDouble) && !asDouble.isInfinite()) {
            value.toLong().toString()
        } else {
            asDouble.toString()
        }
    }
}

File: ios/mithra_flutter_sdk/Sources/mithra_flutter_sdk/NaryaConfigurationCodec.swift
import Foundation
import MithraAnalytics

/// Thrown when the Dart configuration map cannot produce a native configuration.
struct NaryaConfigurationError: Error {
    /// The code reported to Dart, which becomes `NaryaException.code`.
    let code: String
    /// The human-readable reason.
    let message: String
}

/// Decodes the `NaryaConfiguration.toMap()` wire form into the native
/// `Configuration`.
///
/// Defaults mirror the Dart defaults so an absent key and an explicit default
/// behave identically.
enum NaryaConfigurationCodec {

    static func decode(_ source: [String: Any]?) throws -> Configuration {
        guard let source else {
            throw NaryaConfigurationError(
                code: "invalid_argument",
                message: "initialize was called without a configuration."
            )
        }
        guard
            let writeKey = source["writeKey"] as? String,
            !writeKey.isEmpty
        else {
            throw NaryaConfigurationError(
                code: "missing_write_key",
                message: "NaryaConfiguration.writeKey is required and must not be blank."
            )
        }

        let environment: NaryaEnvironment =
            (source["environment"] as? String) == "staging" ? .staging : .production
        let dataPlaneUrl = (source["dataPlaneUrl"] as? String).flatMap {
            $0.isEmpty ? nil : $0
        }

        return Configuration(
            writeKey: writeKey,
            dataPlaneUrl: dataPlaneUrl,
            environment: environment,
            gzipEnabled: source["gzipEnabled"] as? Bool ?? false,
            collectDeviceId: source["collectDeviceId"] as? Bool ?? true,
            trackApplicationLifecycleEvents:
                source["trackApplicationLifecycleEvents"] as? Bool ?? true,
            // Flutter screen tracking is NaryaRouteObserver; the native option
            // would emit one event for the single FlutterViewController, and it
            // also suppresses manual screen() calls.
            trackApplicationScreens: false,
            trackDeepLinks: source["trackDeepLinks"] as? Bool ?? true,
            sessionConfiguration: decodeSession(source["session"] as? [String: Any]),
            inApp: decodeInApp(source["inApp"] as? [String: Any]),
            logLevel: decodeLogLevel(source["logLevel"] as? String)
        )
    }

    /// `SessionConfiguration` does not expose `automaticSessionTracking` and
    /// `sessionTimeoutInMillis` publicly, so the fallbacks come from the SDK's
    /// public default constants instead of an instance.
    private static var defaultAutomaticSessionTracking: Bool {
        Constants.defaultConfig.automaticSessionTrackingStatus
    }
    private static var defaultSessionTimeoutInMillis: UInt64 {
        Constants.defaultConfig.sessionTimeoutInMillis
    }

    private static func decodeSession(_ source: [String: Any]?) -> SessionConfiguration {
        let defaults = SessionConfiguration()
        guard let source else { return defaults }
        let timeout = (source["sessionTimeoutInMillis"] as? NSNumber)?.uint64Value
        return SessionConfiguration(
            automaticSessionTracking: source["automaticSessionTracking"] as? Bool
                ?? defaultAutomaticSessionTracking,
            sessionTimeoutInMillis: timeout ?? defaultSessionTimeoutInMillis,
            updateSessionOnBackgroundEvents:
                source["updateSessionOnBackgroundEvents"] as? Bool
                ?? defaults.updateSessionOnBackgroundEvents
        )
    }

    private static func decodeInApp(_ source: [String: Any]?) -> InAppConfiguration {
        let defaults = InAppConfiguration()
        guard let source else { return defaults }
        let interval = (source["displayIntervalInSeconds"] as? NSNumber)?.doubleValue
        let colorScheme: InAppColorScheme
        switch source["colorScheme"] as? String {
        case "light": colorScheme = .light
        case "dark": colorScheme = .dark
        default: colorScheme = .system
        }
        return InAppConfiguration(
            enabled: source["enabled"] as? Bool ?? defaults.enabled,
            displayInterval: interval ?? defaults.displayInterval,
            autoDisplayPaused: source["autoDisplayPaused"] as? Bool
                ?? defaults.autoDisplayPaused,
            useInMemoryStorage: source["useInMemoryStorage"] as? Bool
                ?? defaults.useInMemoryStorage,
            colorScheme: colorScheme,
            webViewBaseUrl: (source["webViewBaseUrl"] as? String)
                .flatMap { $0.isEmpty ? nil : URL(string: $0) }
        )
    }

    private static func decodeLogLevel(_ source: String?) -> LogLevel {
        switch source {
        case "verbose": return .verbose
        case "debug": return .debug
        case "info": return .info
        case "warn": return .warn
        case "error": return .error
        default: return .none
        }
    }

    /// A comparison key used to decide whether a repeated `initialize` call is
    /// an idempotent no-op or a genuine reconfiguration attempt.
    static func identity(of source: [String: Any]?) -> String {
        guard let source else { return "" }
        return source.keys.sorted()
            .map { "\($0)=\(String(describing: source[$0]))" }
            .joined(separator: "|")
    }
}
